lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -5274,36 +5348,57 @@ ParseResult FIRCircuitParser::parseModule(CircuitOp circuit, bool isPublic,
   return success();
 }
 
+/// formal-old ::= 'formal' id 'of' id ',' 'bound' '=' int info?
+/// formal-new ::= 'formal' id 'of' id ':' info? INDENT (param NEWLINE)* DEDENT
 ParseResult FIRCircuitParser::parseFormal(CircuitOp circuit, unsigned indent) {
   consumeToken(FIRToken::kw_formal);
-  StringRef id, moduleName, boundSpelling;
-  int64_t bound = -1;
+  StringRef id, moduleName;
+  int64_t bound = 0;
   LocWithInfo info(getToken().getLoc(), this);
+  auto builder = circuit.getBodyBuilder();
 
-  // Parse the formal operation
-  if (parseId(id, "expected a formal test name") ||
-      parseToken(FIRToken::kw_of,
-                 "expected keyword 'of' after formal test name") ||
-      parseId(moduleName, "expected the name of a module") ||
-      parseToken(FIRToken::comma, "expected ','") ||
-      parseGetSpelling(boundSpelling) ||
-      parseToken(FIRToken::identifier,
-                 "expected parameter 'bound' after ','") ||
-      parseToken(FIRToken::equal, "expected '=' after 'bound'") ||
-      parseIntLit(bound, "expected integer in bound specification") ||
-      info.parseOptionalInfo())
+  // Parse the name and target module of the formal test.
+  // `formal`
+  if (parseId(id, "expected formal test name") ||
+      parseToken(FIRToken::kw_of, "expected 'of' in formal test") ||
+      parseId(moduleName, "expected module name"))
     return failure();
 
-  // Check that the parameter is valid
-  if (boundSpelling != "bound" || bound <= 0)
-    return emitError("Invalid parameter given to formal test: ")
-               << boundSpelling << " = " << bound,
-           failure();
-
-  // Build out the firrtl mlir op
-  auto builder = circuit.getBodyBuilder();
-  builder.create<firrtl::FormalOp>(info.getLoc(), id, moduleName, bound);
+  // TODO: Remove the old `, bound = N` variant in favor of the new parameters.
```

Comment:
Does Chisel emit the old `bound = N` currently? It's unfortunate that we have to keep old style as well 😿

Reply:
It's only until the thing is in the spec. I'll drop this again afterwards. Chisel doesn't emit anything as of today 😁